### DISCUSSION TITLE
Is there a way to change venv without restarting FastAPI server?

### DISCUSSION THREAD

**Discussion by seunggs** [fastapi/fastapi#9643] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
N/A
```

### Description

Hi,

I have a use case where I need to change venv without restarting FastAPI server - is this at all possible? 

I tried to add the venv dir to PATH manually in the code, but that does not work. As a last resort, I'm running a subprocess to run the function using a different venv, but this adds 4-6 seconds of latency to every call I need to make.

Any help here would be greatly appreciated - thank you!

### Operating System

Linux

### FastAPI Version

0.92.0

### Python Version

3.10.11

**Comment by jgould22** [comment]
The short answer is not really .

Why is it you want want to change the venv? Are you using 2 different versions of the same library ?

**Reply by seunggs** [reply]
Hi, thanks for the quick response. 

No, actually I have a use case where I need to parse the user code to extract the functions and get its args, returns, and their types. I couldn't find a way to do that reliably via AST, so I'm dynamically importing the user code via `importlib` (`exec_module`) so that I can use utils like `typing.get_type_hints`. 

Since our users can have multiple projects each with its own `venv`, we need a way to run this dynamic import in a venv (otherwise, it would throw "module not found" error since FastApi is running in a different `venv`), but I couldn't find a way to do that. 

I figured if there's a way to change `venv` before running this dynamic import code, it would solve my problem, but doesn't look like there's a way to do this without restarting FastAPI server. It works if I spawn a subprocess under the correct `venv` of course, but since this call is made frequently, the 4-6 sec latency due to spawning a subprocess everytime is pretty damaging.

This is outside the scope of FastAPI of course, but if you have any other ideas to do this, I'd appreciate the input. Thanks!

**Reply by jgould22** [reply]
Ah I am not sure I have much to add.

If I was running user code like that with their own venv I might try and spin up light weight vm (like firecracker) or container with an agent inside which could make subsequent calls fast during the users session. It all really depends on what your service usage looks like though and what you are ultimately trying to achieve. 

Best of luck.

**Reply by seunggs** [reply]
Got it thank you
